select fast solution based on Number rather than Symbol

smichr opened the pull request:
The intent of the code selecting a fast path for a matrix
inversion checked for Symbols to detect non-Numeric matrices.
A more successful check is to check for Numbers.

Fixes #12448

smichr committed:
select fast solution based on Number rather than Symbol

The intent of the code selecting a fast path for matrix
solution checked for Symbols to detect non-Numeric matrices.
A more successful check is to check for Numbers.

asmeurer commented on sympy/solvers/tests/test_solvers.py:
```diff
@@ -1845,3 +1845,25 @@ def test_inf():
+    ts = time() - t
+    # the two solutions should be found in nearly the same amount
+    # of time.
+    assert min(tf, ts)/max(tf, ts) > 0.9
```
I am a little worried about tests based on timings, but since we are testing ratios this should pass even on a really slow machine.

smichr replied:
agreed on both accounts

asmeurer replied:
It looks like this test failed on Travis.

smichr replied:
And guess why:

```
_____________ sympy/solvers/tests/test_solvers.py:test_issue_12448 _____________
  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/sympy/solvers/tests/test_solvers.py", line 1869, in test_issue_12448
    assert min(tf, ts)/max(tf, ts) > 0.9
AssertionError
```

Trust but verify! ;-)

smichr replied:
Instead of testing that, perhaps I can test that the output is consistent with the path that it went through.

smichr replied:
Interesting result: the function form, rather than the symbol form, is faster by almost a factor of 3 whether it is run first or second.  (Relative times were 2.3 vs 6.5, a factor of 2.8.)

asmeurer replied:
Possibly caching. I'm sure the symbols x:15 and subexpressions aren't in the cache before this function is run.

smichr committed:
12448: make test more robust

Regarding the test added:

    The soutions should be the same if they go through the same
    solution path. Although the two solutions differ in terms of
    how a minus sign is distributed, the op count is the same.
    The op count is significantly different if two different paths
    are taken when solving the system of equations. The times
    taken to solve each system should be similar, too. Before this
    PR, the time to solve the 3 eqn system in terms of functions took
    about 20 times longer to solve than the system written in terms
    of symbols.
